### PR TITLE
ZMS-85

### PR DESCRIPTION
**Pull request by NickOvt**
The main idea was to add functionality that will generate a `markdown` file with a table that has overview about all the API endpoints and their tests. So the table will have the api path the method of the path the amount of tests for the path also the amount of positive tests and negative tests for the path.

1. Added `/api-methods/:arg` ( `:arg` is not used) GET endpoint to the API. Used only in `test` environment and fetches all the available endpoints that the API has.
2. in the `test` folder added `_globals-test.js` file that has a global `beforeEach()` and `after()` mocha hooks (global in this context means that `beforeEach()` is ran before every test in the test run, same for `after()`, it runs after all the tests in the given test run). `beforeEach()` checks that the given test `title` contains at least an api method in it and an expectation (`success` or `failure`) then adds the `title` to a `titles` array.
`after()` hook:
- calls the `/api-method` to retrieve an object with all api endpoints and their corresponding data
- loop through the `routes` and construct a map with a structure:
```javascript
{
[method_path] = {
      method: "post" | "put" | "delete" | "get",
      path: "api path",
      name: "name of the path if given",
      testCount: number,
      positiveTestCount: number,
      negativeTestCount: number,
  }
}
``` 
- loop through ran test titles and parse them, retrieve `method`, `path`, `expectedResult` (`success` or `failure`) data from the title
- get corresponding data from the map constructed before
- if there is data about given path and method then if the test is positive (`expectedResult === "success"`) increment `positiveTestCount` if the test was negative (`expectedResult === "failure"`) increment `negativeTestCount`. In either case increment general `testCount` as well.
- sort final data. First by `testCount` then the `path` alphabetically
- construct a string for the table for a `markdown` file
- write data to `api-tests-overview.md` file
- also log any test titles that did not pass checks
3. Updated all tests in the `/test` folder to accomodate new generating script

### CODE REVIEW

File: test/_globals-test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    if (/POST|PUT|DELETE|GET|post|put|delete|get/.test(currentTestTitle) && /success|failure/.test(currentTestTitle)) {
+        titles.push(currentTestTitle);
+    } else {
+        unsupportedTitles.push(currentTestTitle);
+    }
+});
+
+// eslint-disable-next-line no-undef
+after("Generate test overview table after all tests", async () => {
+    try {
```
Comment (andris9) [inline test/_globals-test.js:25]: Maybe not use `try..catch` here. If there is an exception, the `after()` handler will catch it and log it as a failure.

File: api.js [hunk 2]
```diff
@@ -562,6 +562,22 @@ module.exports = done => {
     webhooksRoutes(db, server);
     settingsRoutes(db, server, settingsHandler);
 
+    if (process.env.NODE_ENV === "test") {
+        server.get(
+            { name: 'api-methods', path: '/api-methods/:arg' /*, someKey: Joi.object({arg: Joi.string()})*/ },
+            tools.responseWrapper(async (req, res) => {
+                res.charSet('utf-8');
+    
+                // const obj = Object.assign({name: "initial name"}, server.router.getRoutes().getsettingskey.spec);
```
Comment (andris9) [inline api.js:571]: Remove unneeded comments, no reason to keep these in the code

File: test/_globals-test.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        for (const routeName in routes) {
+            const route = routes[routeName];
+            const method = route.spec.method;
+            const path = route.spec.path;
+
+            mapApiMethodToSpec[`${method.toLowerCase()}_${path}`] = {method, path, name: route.spec.name || route.name, testCount: 0, positiveTestCount: 0, negativeTestCount: 0};
+        }
+
+        for (const title of titles) {
+            const titleSplit = title.split(" ");
```
Comment (andris9) [inline test/_globals-test.js:43]: It is safer to use a regex for splitting.
```
title.split(/\s+/);
```

File: test/_globals-test.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            const method = titleSplit[1].toLowerCase();
+            const path = titleSplit[2];
+            const expectedResult = titleSplit[4];
+
+            // missing method or path (string is too short to be accepted as valid test title)
+            if (!method || !path) {
+                continue;
+            }
+
+            const data = mapApiMethodToSpec[`${method.toLowerCase()}_${path.replace(/{/g, ":").replace(/}/g, "")}`];
```
Comment (andris9) [inline test/_globals-test.js:53]: `method` is already converted to lowercase

File: test/_globals-test.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+            // wront path or wrong data etc. (no such route, can't construct route from test title)
```
Comment (andris9) [inline test/_globals-test.js:55]: Typo in comment

File: test/_globals-test.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            } else {
+                return 0;
+            }
+        });
+
+        for (const data of sortedData) {
+            content += `| \`${data.path}\` | \`${data.method}\` | ${data.testCount} | ${data.positiveTestCount > 0 ? "✅" : "❌"} (${data.positiveTestCount}) | ${data.negativeTestCount > 0 ? "✅" : "❌"} (${data.negativeTestCount}) |\n`;
+        }
+
+        fs.writeFile("./api-tests-overview.md", content, (err) => {
```
Comment (andris9) [inline test/_globals-test.js:91]: Wait until the file is written.
```
await fs.promises.writeFile('./api-tests-overview.md', content);
```

Comment (andris9) [inline test/_globals-test.js:91]: Prefer absolute path names with file paths, eg. `__dirname+'/../api-tests-overview.md'` as no-one knows from where the `npm test` is called and what is the actual working directory.

File: test/_globals-test.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
+'use strict'
+
+const supertest = require('supertest');
+const fs = require("fs");
```
[the remark this answers is not in the document]
Comment (NickOvt) [reply test/_globals-test.js]: Added prettier to vscode and resaved the file. Fixed style.

File: api.js [hunk 8]
```diff
@@ -562,6 +562,22 @@ module.exports = done => {
     webhooksRoutes(db, server);
     settingsRoutes(db, server, settingsHandler);
 
+    if (process.env.NODE_ENV === "test") {
+        server.get(
+            { name: 'api-methods', path: '/api-methods/:arg' /*, someKey: Joi.object({arg: Joi.string()})*/ },
```
[the remark this answers is not in the document]
Comment (NickOvt) [reply api.js]: removed the :arg part

File: api.js (hunk 2 above)
Comment (NickOvt) [reply api.js]: removed the unneeded comments

File: test/_globals-test.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+const server = supertest.agent(`http://127.0.0.1:${config.api.port}`);
+
+const titles = [];
+const unsupportedTitles = [];
+
+// global beforeEach to run before EVERY test
+beforeEach("Get test data before each test", async function () {
+    const currentTestTitle = this.test.ctx.currentTest.title; // eslint-disable-line no-invalid-this
+    if (/POST|PUT|DELETE|GET|post|put|delete|get/.test(currentTestTitle) && /success|failure/.test(currentTestTitle)) {
```
[the remark this answers is not in the document]
Comment (NickOvt) [reply test/_globals-test.js]: Indeed, could've just make it case insensitive :).
Fixed

File: test/_globals-test.js (hunk 1 above)
Comment (NickOvt) [reply test/_globals-test.js]: removed try catch

File: test/_globals-test.js (hunk 6 above)
Comment (NickOvt) [reply test/_globals-test.js]: whoops :D, fixed

File: test/_globals-test.js (hunk 5 above)
Comment (NickOvt) [reply test/_globals-test.js]: whoops :D, fixed

File: test/_globals-test.js (hunk 6 above)
Comment (NickOvt) [reply test/_globals-test.js]: rewrote to use fs.promises. fixed

File: test/_globals-test.js (hunk 3 above)
Comment (NickOvt) [reply test/_globals-test.js]: done

File: test/_globals-test.js (hunk 6 above)
Comment (NickOvt) [reply test/_globals-test.js]: Okay will do with `__dirname`. Added `__dirname` and path. Fixed

**Review by andris9 — APPROVED** [review]
